darts/models/regression_model.py

```diff
@@ -13,300 +13,259 @@
+            X=self.train_x,
+            y=self.train_y,
+            **kwargs
+        )
+        if self.max_lag == 0:
+            self.prediction_data = pd.DataFrame(columns=series.stack(other=exog).columns())
+        elif exog is not None:
+            self.prediction_data = series.stack(other=exog)[-self.max_lag:]
+        else:
+            self.prediction_data = series[-self.max_lag:]
 
-        raise_if_not(len(train_features) > 0, 'Need at least one feature series', logger)
-        raise_if_not(all([s.has_same_time_as(train_target) for s in train_features]),
-                     'All provided time series must have the same time index', logger)
-        self.train_features = train_features
-        self.train_target = train_target
         self._fit_called = True
 
-    @abstractmethod
-    def predict(self, features: List[TimeSeries]) -> TimeSeries:
-        """ Predicts values of the target time series, given a list of features time series
+    def _create_training_data(self, series: TimeSeries, exog: TimeSeries = None):
+        """ Create dataframe of exogenous and endogenous variables containing lagged values.
 
         Parameters
         ----------
-        features
-            The list of features time series, of the same length
+        series : TimeSeries
+            Target series.
+        exog : TimeSeries, optional
+            Exogenous variables.
 
         Returns
         -------
-        TimeSeries
-            A series containing the predicted targets, of the same length as the features series
+        pd.dataframe
+            Data frame with lagged values.
         """
+        raise_if(series.width > 1,
+            "Series must not be multivariate. Pass exogenous variables to 'exog' parameter.",
+            logger
+        )
+        training_data = series.pd_dataframe()
 
-        if (not self._fit_called):
-            raise_log(Exception('fit() must be called before predict()'), logger)
-
-        length_ok = len(features) == len(self.train_features)
-        dimensions_ok = all(features[i].width == self.train_features[i].width for i in range(len(features)))
-        raise_if_not(length_ok and dimensions_ok,
-                     'The number and dimensionalities of all given features must correspond to those used for'
-                     ' training.', logger)
+        if self.lags is not None:
+            lagged_data = self._create_lagged_data(series=series, lags=self.lags)
+            training_data = pd.concat([training_data, lagged_data], axis=1)
 
-    def _historical_forecasts_sanity_checks(self, *args: Any, **kwargs: Any) -> None:
-        """Sanity checks for the historical_forecasts function
+        if self.lags_exog is not None:
+            for col in exog.columns():
+                lagged_data = self._create_lagged_data(
+                    series=exog[col], lags=self.lags_exog
+                )
+                training_data = pd.concat([training_data, lagged_data], axis=1)
```

Comment:
Not sure if I'm right, but it could be that calling `pd.concat()` repeatedly is less efficient than just storing the dataframes in a list and calling `pd.concat()` once at the end on the whole list.

Reply:
Wow, yes you're right here. Just did some simple tests and it really makes a large difference even for smaller datasets. Changed it and I'll keep that in mind.